### PR TITLE
Expand PhotonLib and PhotonTracking functionalities

### PR DESCRIPTION
**Pull request by srimanachanta**
As more and more new features are added to PhotonVision, the robot code client lib seems to be feeling a little lackluster. I wanted to create this PR in order to add some ease-of-life features to PhotonLib. Here are some of the main things I want to add:
- Clean up some of the extra or useless structurings within the classes, such as excessive publicly exposed variables.
- In order to integrate with utilities within the camera, add a parameter to the PhotonCamera constructor to accept a "mount config" of the camera in the RCS like [this](https://github.com/Talon540Programming/TalonLib/blob/068680ae94adc25c296e475ecf7c3f4902775b69/src/main/java/org/talon540/sensors/vision/VisionCameraMountConfig.java).
- As previously stated, integrate a lot of the functionalities from the PhotonUtils class and other similar classes from wpi directly into the camera using a system of Optional return methods. By doing this, we can further remove a lot of the hassle behind the calculations from the user. When #571 is merged, this can also be included.
- Move sim classes into their own sim folder as done in WPI.

### CODE REVIEW

File: photon-lib/src/main/java/org/photonvision/PhotonCamera.java [hunk 1]
```diff
@@ -55,11 +63,11 @@ public class PhotonCamera {
...
-    StringSubscriber versionEntry;
-    BooleanPublisher inputSaveImgEntry, outputSaveImgEntry;
-    IntegerEntry pipelineIndexEntry, ledModeEntry;
-    IntegerSubscriber heartbeatEntry;
 
+    /**
+     * Close out the NetworkTables entries used to communicate with the camera and releases their
+     * resources.
+     */
     public void close() {
```
Comment (srimanachanta) [inline photon-lib/src/main/java/org/photonvision/PhotonCamera.java:71]: what is the point of this method, most teams create the object in RobotInit and then leave the object for the lifecycle of the robot code, this serves no purpose as far as I can tell.

File: photon-lib/src/main/java/org/photonvision/PhotonCamera.java [hunk 2]
```diff
@@ -43,9 +43,17 @@
...
+    // NetworkTables entries
+    // Used
     RawSubscriber rawBytesEntry;
     BooleanEntry driverModeEntry;
+    StringSubscriber versionEntry;
+    BooleanPublisher inputSaveImgEntry, outputSaveImgEntry;
+    IntegerEntry pipelineIndexEntry, ledModeEntry;
+    IntegerSubscriber heartbeatEntry;
+
     BooleanPublisher driverModePublisher;
```
Comment (srimanachanta) [inline photon-lib/src/main/java/org/photonvision/PhotonCamera.java:57]: What are these publishers, subscribers, and entires for? They are uninitiated and unused by any file as far as I can tell.

File: photon-lib/src/main/java/org/photonvision/PhotonCamera.java [hunk 3]
```diff
@@ -43,9 +43,17 @@
 
 /** Represents a camera that is connected to PhotonVision. */
 public class PhotonCamera {
-    protected final NetworkTable rootTable;
+    protected final NetworkTable cameraTable;
+
+    // NetworkTables entries
+    // Used
     RawSubscriber rawBytesEntry;
```
Comment (srimanachanta) [inline photon-lib/src/main/java/org/photonvision/PhotonCamera.java:50]: Why is this public? Is the purpose to expose the entries to the user for them to use in their own code?

File: photon-lib/src/main/java/org/photonvision/PhotonCamera.java (hunk 2 above)
Comment (mcm001) [reply photon-lib/src/main/java/org/photonvision/PhotonCamera.java:57]: heartbeatEntry, pipelineIndexEntry, ledModeEntry, inputSaveImgEntry/outputSaveImgEntry and versionentry are all used....

File: photon-lib/src/main/java/org/photonvision/PhotonCamera.java (hunk 1 above)
Comment (mcm001) [reply photon-lib/src/main/java/org/photonvision/PhotonCamera.java:71]: PhotonCamera should probably implement AutoCloseable, and we should make sure all the publishers/subscribers/entries are close()d. it's for NT JNI resource cleanup.

File: photon-lib/src/main/java/org/photonvision/PhotonCamera.java (hunk 2 above)
Comment (srimanachanta) [reply photon-lib/src/main/java/org/photonvision/PhotonCamera.java:57]: Yes, im referring to those bellow, the ones I tagged with `// unused`

File: photon-lib/src/main/java/org/photonvision/PhotonCamera.java (hunk 3 above)
Comment (mcm001) [reply photon-lib/src/main/java/org/photonvision/PhotonCamera.java:50]: They are not public, they're protected? It's so SimPhotonCamera can access them.

File: photon-lib/src/main/java/org/photonvision/PhotonCamera.java (hunk 1 above)
Comment (srimanachanta) [reply photon-lib/src/main/java/org/photonvision/PhotonCamera.java:71]: Ok, I can add that then.

File: photon-lib/src/main/java/org/photonvision/PhotonCamera.java (hunk 3 above)
Comment (srimanachanta) [reply photon-lib/src/main/java/org/photonvision/PhotonCamera.java:50]: Java sets those to public when not explicitly stated unless I am remembering incorrectly

Comment (mcm001) [reply photon-lib/src/main/java/org/photonvision/PhotonCamera.java:50]: Nope. Default is package private.

**Comment by gerth2** [comment]
Will do a full review later this week or early next if no one beats me to it - but I do wanna say I do like the mounting configuration utiltiy class.

Reorganizing the sim classes is also probably fine. Big thing is if package imports change, want to make sure that doesn't break example code on people. I'd say this one either needs to sneak in before beta 8, or wait till next season just to avoid breaking end users.

**Comment by mcm001** [comment]
I think we'll punt this one to 2024?

**Comment by srimanachanta** [comment]
> I think we'll punt this one to 2024?

Yes, most of this realizes on the hardware config class and that is something I would consider breaking as people will need to adapt their code around it

**Pull request closed without merging**